### DISCUSSION TITLE
MSI package UI shows "Install" upon uninstallation

### DISCUSSION THREAD

**Discussion by mindonwarp** [wixtoolset/issues#7977] [Questions]
Hi everyone,

 I've noticed that the MSI package's UI has "Install" as caption on the "Install" button even during uninstallation. When I open up the "installed apps" and right-click -> uninstall on a program installed with a WiX bundle, the first dialog correctly shows "Uninstall", but when I click on uninstall the next dialog that pops up (the one from the MSI) looks exactly like the install version of it.

The UI wxs is the same one from github (WixUI_InstallDir), minus the License Agreement.

What am I doing wrong?
Thanks

**Comment by ag-ctlg** [comment]
Any progress on this?

**Comment by bevanweiss** [comment]
You're probably going to have to provide a screenshot showing which button you mean.
And the wxs that you're using for your dialog controls.  I'd say the button on that dialog isn't conditional around Install / Uninstall.

Since you mention 'Install' then it's probably the English language that you're using, so the various strings can be found here:
https://github.com/wixtoolset/wix/blob/main/src/ext/UI/wixlib/WixUI_en-us.wxl

**Comment by ag-ctlg** [comment]
Hi, thanks for the reply! Here are the screenshots and the wxs file:

![2024-07-11_14h51_13](https://github.com/wixtoolset/issues/assets/99892060/aee15306-2914-42bb-b0cc-f78458dbc26b)
![2024-07-11_14h51_35](https://github.com/wixtoolset/issues/assets/99892060/4496d580-e0dd-4494-bf01-a79c4da2738e)

<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
   <Fragment>
      <UI Id="WixUI_Custom">
         <TextStyle Id="WixUI_Font_Normal" FaceName="Tahoma" Size="8" />
         <TextStyle Id="WixUI_Font_Bigger" FaceName="Tahoma" Size="12" />
         <TextStyle Id="WixUI_Font_Title" FaceName="Tahoma" Size="9" Bold="yes" />

         <Property Id="DefaultUIFont" Value="WixUI_Font_Normal" />
         <Property Id="WixUI_Mode" Value="InstallDir" />

         <DialogRef Id="BrowseDlg" />
         <DialogRef Id="DiskCostDlg" />
         <DialogRef Id="ErrorDlg" />
         <DialogRef Id="FatalError" />
         <DialogRef Id="FilesInUse" />
         <DialogRef Id="MsiRMFilesInUse" />
         <DialogRef Id="PrepareDlg" />
         <DialogRef Id="ProgressDlg" />
         <DialogRef Id="ResumeDlg" />
         <DialogRef Id="UserExit" />

         <Publish Dialog="BrowseDlg" Control="OK" Event="DoAction" Value="WixUIValidatePath_$(sys.BUILDARCHSHORT)" Order="3" />
         <Publish Dialog="BrowseDlg" Control="OK" Event="SpawnDialog" Value="InvalidDirDlg" Order="4" Condition="NOT WIXUI_DONTVALIDATEPATH AND WIXUI_INSTALLDIR_VALID&lt;&gt;&quot;1&quot;" />

         <Publish Dialog="ExitDialog" Control="Finish" Event="EndDialog" Value="Return" Order="999" />

         <Publish Dialog="WelcomeDlg" Control="Next" Event="NewDialog" Value="InstallDirDlg" Condition="NOT Installed" />
         <Publish Dialog="WelcomeDlg" Control="Next" Event="NewDialog" Value="VerifyReadyDlg" Condition="Installed AND PATCH" />

         <Publish Dialog="InstallDirDlg" Control="Back" Event="NewDialog" Value="WelcomeDlg" />
         <Publish Dialog="InstallDirDlg" Control="Next" Event="SetTargetPath" Value="[WIXUI_INSTALLDIR]" Order="1" />
         <Publish Dialog="InstallDirDlg" Control="Next" Event="DoAction" Value="WixUIValidatePath_$(sys.BUILDARCHSHORT)" Order="2" Condition="NOT WIXUI_DONTVALIDATEPATH" />
         <Publish Dialog="InstallDirDlg" Control="Next" Event="SpawnDialog" Value="InvalidDirDlg" Order="3" Condition="NOT WIXUI_DONTVALIDATEPATH AND WIXUI_INSTALLDIR_VALID&lt;&gt;&quot;1&quot;" />
         <Publish Dialog="InstallDirDlg" Control="Next" Event="NewDialog" Value="VerifyReadyDlg" Order="4" Condition="WIXUI_DONTVALIDATEPATH OR WIXUI_INSTALLDIR_VALID=&quot;1&quot;" />
         <Publish Dialog="InstallDirDlg" Control="ChangeFolder" Property="_BrowseProperty" Value="[WIXUI_INSTALLDIR]" Order="1" />
         <Publish Dialog="InstallDirDlg" Control="ChangeFolder" Event="SpawnDialog" Value="BrowseDlg" Order="2" />

         <Publish Dialog="VerifyReadyDlg" Control="Back" Event="NewDialog" Value="InstallDirDlg" Order="1" Condition="NOT Installed" />
         <Publish Dialog="VerifyReadyDlg" Control="Back" Event="NewDialog" Value="MaintenanceTypeDlg" Order="2" Condition="Installed AND NOT PATCH" />
         <Publish Dialog="VerifyReadyDlg" Control="Back" Event="NewDialog" Value="WelcomeDlg" Order="2" Condition="Installed AND PATCH" />

         <Publish Dialog="MaintenanceWelcomeDlg" Control="Next" Event="NewDialog" Value="MaintenanceTypeDlg" />

         <Publish Dialog="MaintenanceTypeDlg" Control="RepairButton" Event="NewDialog" Value="VerifyReadyDlg" />
         <Publish Dialog="MaintenanceTypeDlg" Control="RemoveButton" Event="NewDialog" Value="VerifyReadyDlg" />
         <Publish Dialog="MaintenanceTypeDlg" Control="Back" Event="NewDialog" Value="MaintenanceWelcomeDlg" />

         <Property Id="ARPNOMODIFY" Value="1" />
      </UI>

      <UIRef Id="WixUI_Common" />
   </Fragment>
</Wix>

**Reply by bevanweiss** [reply]
That looks like the Resume Dialog, it's more for if a setup was 'interrupted', like if a restart was needed or something similar.

I would have expected on Uninstall for it to enter at the `MaintenanceWelcomeDlg`, and then to proceed to the `MaintenanceTypeDlg` before then proceeding to the `VerifyReadyDlg`.


What version of the WiX Toolset are you using?
If you have either `VerifyReadyDlg.wxs` or `ResumeDlg.wxs` files in your project, you should paste them here.

**Comment by ag-ctlg** [comment]
I would expect the same and that's what confuses me. WiX version is 4.0.5 at the moment.
We don't have any other custom UI wxs files.

When I look at the log file, the first few lines look like this:

MSI (c) (F0:DC) [07:59:30:926]: Note: 1: 2205 2:  3: PatchPackage
MSI (c) (F0:DC) [07:59:30:926]: Machine policy value 'DisableRollback' is 0
MSI (c) (F0:DC) [07:59:30:926]: User policy value 'DisableRollback' is 0
MSI (c) (F0:DC) [07:59:30:926]: PROPERTY CHANGE: Adding UILevel property. Its value is '5'.
MSI (c) (F0:DC) [07:59:30:926]: Note: 1: 2203 2: C:\Windows\Installer\inprogressinstallinfo.ipi 3: -2147287038 
MSI (c) (F0:DC) [07:59:30:926]: PROPERTY CHANGE: Adding Preselected property. Its value is '1'.
MSI (c) (F0:DC) [07:59:30:926]: APPCOMPAT: [DetectVersionLaunchCondition] Launch condition already passes.
**MSI (c) (F0:DC) [07:59:30:926]: PROPERTY CHANGE: Adding ACTION property. Its value is 'INSTALL'.**
MSI (c) (F0:DC) [07:59:30:926]: Doing action: INSTALL
Action 7:59:30: INSTALL. 
Action start 7:59:30: INSTALL.
MSI (c) (F0:DC) [07:59:30:926]: UI Sequence table 'InstallUISequence' is present and populated.
MSI (c) (F0:DC) [07:59:30:926]: Running UISequence
MSI (c) (F0:DC) [07:59:30:926]: PROPERTY CHANGE: Adding EXECUTEACTION property. Its value is 'INSTALL'.

Is it normal for it to start the INSTALL action? Later on in the log I see the ResumeDlg starting.
It does perform an uninstall in the end, but the dialog shows "Install".

Oh and btw, the process starts from the setup bundle, not the MSI. When I start the MSI I get the "change/repair/remove" dialog.
However, the MSI is never shipped alone because there are prerequisites that need to be pre-installed, thus the bundle.

**Reply by ahodzic80** [reply]
Did you ever get this resolved? I am also facing the same problem.

**Reply by ag-ctlg** [reply]
> Did you ever get this resolved? I am also facing the same problem.

No, still open.

**Reply by bevan-howie** [reply]
Same problem here.

**Comment by HubbaBob** [comment]
We have the exact same problem: Setupbundle installing an MSI (which installs a service)
When running uninstall from add/remove or direclty from the bundle it shows the resume/install dialog.  Continuing anyway actually uninstalls the application, but the install dlgs are very confusing.

Using wix 4.0.5

**Comment by HubbaBob** [comment]
Did some more investigating, the problem persists when using wix5.

However, compared behaviour to an older wix installer bundle build with wix3. The old way DOES NOT show the InternalUI of the MSI when uninstalling via bundle, even with DisplayInternalUI="yes". So even if there is the same problem in wix3 it does not show because of that.

As a quick fix one can mimick the wix3 behaviour by disabling the internalUI of the MSI when uninstalling, like so:

`<MsiPackage Id="MSIPackage" SourceFile="acme.msi" Vital="yes" bal:DisplayInternalUICondition="WixBundleAction &lt;&gt; 4">`

It seems the problem was introduced when switching from
`DisplayInternalUI="yes"`
to
`bal:DisplayInternalUICondition="1"`

However I am not sure if this intended behaviour or a bug. Maybe someone from the team can comment?

Thx!

**Reply by bevan-howie** [reply]
@HubbaBob your proposed workaround works a treat - thanks heaps!

**Reply by bevanweiss** [reply]
> Did some more investigating, the problem persists when using wix5.
> 
> However, compared behaviour to an older wix installer bundle build with wix3. The old way DOES NOT show the InternalUI of the MSI when uninstalling via bundle, even with DisplayInternalUI="yes". So even if there is the same problem in wix3 it does not show because of that.
> 
> As a quick fix one can mimick the wix3 behaviour by disabling the internalUI of the MSI when uninstalling, like so:
> 
> `<MsiPackage Id="MSIPackage" SourceFile="acme.msi" Vital="yes" bal:DisplayInternalUICondition="WixBundleAction &lt;&gt; 4">`
> 
> It seems the problem was introduced when switching from `DisplayInternalUI="yes"` to `bal:DisplayInternalUICondition="1"`
> 
> However I am not sure if this intended behaviour or a bug. Maybe someone from the team can comment?
> 
> Thx!

It does look like WiX v3 specifically didn't show InternalUI during uninstall / repair

https://github.com/wixtoolset/wix3/blob/b40e9a32c24033e11b77baf2c91a704382f898ed/src/burn/engine/msiengine.cpp#L1352-L1353
```csharp
    // suppress internal UI during uninstall to mimic ARP and "msiexec /x" behavior
    if (fDisplayInternalUI && BOOTSTRAPPER_ACTION_STATE_UNINSTALL != actionState && BOOTSTRAPPER_ACTION_STATE_REPAIR != actionState)
```

Potentially this similar conditional of actionState != ..UNINSTALL/REPAIR could be added to the `fShowInternalUI` test.
https://github.com/wixtoolset/wix/blob/f440fb317c630e7bc6d4ee4d657a200654e2f876/src/ext/Bal/stdbas/WixStandardBootstrapperApplication.cpp#L672-L675
although the argument could also be made that with the revised conditional capability of the WiX v4+ DisplayInternalUICondition that it could be left as an author consideration whether to include it or not.
Although I'd argue the default upgrade conversion should then include these action conditions to maintain the pre WiX v4 behaviour following migration.

**Reply by melvinabjohn** [reply]
@HubbaBob Amazing..!! I spend so much time looking for a fix changing what all..only God knows. I upgraded from wix 3 to v4(Wix5) and started getting this issue.

**Comment by ag-ctlg** [comment]
@HubbaBob thanks for the workaround!

But still, this is a workaround, right? The internal UI doesn't get displayed so the user doesn't see the word "Install" written on a button where they expect to see "Uninstall". But getting the button to actually show the correct text is not possible or what's the issue behind that?

Should an issue be created from this discussion or should we accept the workaround as the solution?

**Reply by HubbaBob** [reply]
Just checked this again. It seems that this is wanted and documented behaviour. There is an old issue about this

https://github.com/wixtoolset/issues/issues/8695

which links to the documentation

https://wixtoolset.org/docs/fourthree/faqs/#removal-of-displayinternalui

Seems this was introduced to give you more control over the InternalUI, the proposed way in the FAQ to get Wix3 behaviour is

`bal:DisplayInternalUICondition="WixBundleAction = 6"`

So I guess that's the accepted way and we can close this discussion.

**Reply by ag-ctlg** [reply]
OK, sure, concerning the display of the internal UI of bundled MSI packages, this was a deliberate change which is also documented in the migration FAQ. 👍 

I can live with not displaying the internal UI when uninstalling, since there is nothing special to be handled inside the MSI package during uninstall.

But what if you want to show the internal UI for some reason? 
Is there (or should there be) some sort of a parameter that should be sent from the bundle to the packaged MSI, so that it knows it's running in uninstall mode?
